rest/method_item_put_test.go

```diff
@@ -294,6 +363,179 @@ func TestPutItem(t *testing.T) {
 			ResponseBody: `{"id": "1", "foo": "3"}`,
 			ExtraTest:    checkPayload("foo.sub", "1", map[string]interface{}{"id": "1", "foo": "3"}),
 		},
+
+		`pathID:not-found,body:valid,default:missing`: {
+			Init: initForDefaultField(now, yesterday),
+			NewRequest: func() (*http.Request, error) {
+				body := bytes.NewReader([]byte(`{"foo": "baz"}`))
+				return http.NewRequest("PUT", "/foo/66", body)
+			},
+			ResponseCode: http.StatusCreated,
+			ResponseBody: `{"id": "66", "foo": "baz", "bar": "default"}`,
+			ExtraTest:    checkPayload("foo", "66", map[string]interface{}{"id": "66", "foo": "baz", "bar": "default"}),
+		},
+		`pathID:not-found,body:valid,default:set`: {
+			Init: initForDefaultField(now, yesterday),
+			NewRequest: func() (*http.Request, error) {
+				body := bytes.NewReader([]byte(`{"foo": "baz", "bar": "value"}`))
+				return http.NewRequest("PUT", "/foo/66", body)
+			},
+			ResponseCode: http.StatusCreated,
+			ResponseBody: `{"id": "66", "foo": "baz", "bar": "value"}`,
+			ExtraTest:    checkPayload("foo", "66", map[string]interface{}{"id": "66", "foo": "baz", "bar": "value"}),
+		},
+		`pathID:found,body:valid,default:missing`: {
+			Init: initForDefaultField(now, yesterday),
+			NewRequest: func() (*http.Request, error) {
+				body := bytes.NewReader([]byte(`{"foo": "baz"}`))
+				return http.NewRequest("PUT", "/foo/1", body)
+			},
+			ResponseCode: http.StatusOK,
+			ResponseBody: `{"id": "1", "foo": "baz"}`,
+			ExtraTest:    checkPayload("foo", "1", map[string]interface{}{"id": "1", "foo": "baz"}),
+		},
+		`pathID:found,body:valid,default:set`: {
+			Init: initForDefaultField(now, yesterday),
+			NewRequest: func() (*http.Request, error) {
+				body := bytes.NewReader([]byte(`{"foo": "baz", "bar": "value"}`))
+				return http.NewRequest("PUT", "/foo/1", body)
+			},
+			ResponseCode: http.StatusOK,
+			ResponseBody: `{"id": "1", "foo": "baz", "bar": "value"}`,
+			ExtraTest:    checkPayload("foo", "1", map[string]interface{}{"id": "1", "foo": "baz", "bar": "value"}),
+		},
+		`pathID:found,body:valid,default:delete`: {
+			Init: initForDefaultField(now, yesterday),
+			NewRequest: func() (*http.Request, error) {
+				body := bytes.NewReader([]byte(`{"foo": "baz"}`))
+				return http.NewRequest("PUT", "/foo/2", body)
+			},
+			ResponseCode: http.StatusOK,
+			ResponseBody: `{"id": "2", "foo": "baz"}`,
```

Comment:
I think here we should not clear  the `bar` value.

Consider this case:

A "home automation" application starts up by reading a list of "lights" from a config file, and it PUTs all the lights,so that it can later rely on them being there. We want the app to use the server defaults when there are no value configured in the config file. The first time we run the app, all the lights end up with the correct values, but whn we restart it, and rerun all the PUT requests, we will (unintentionally) end up clearing all non-required defaults.

PUT is meant to hold _replace_ semantics, so ideally, I think resetting the values to the _default_ is the correct thing to do, but since we won't/can't do that for read-only fields (e.g. createdTime), I can accept not doing it for _default_ fields neither. However, _clear_ fields with server-side defaults, no matter if they are required or not, may have odd side-effects in real applications as calling N PUT request has a semantically very different effect compared to calling just one.

Reply:
Yes, I hear you. But then, what is the difference between (Default) and (Required,Default)? I guess (Default) can be explicitly deleted with PATCH(JSON-Patch), wheres (Required,Default) will get its Default value on delete?

Reply:
I am puzzled how we should proceed with (Default) in this case. On one hand PUT is used to "PUT exactly this resource there!", then we set some fields under its feet, and for (Default) fields we can loose ability to delete them.

Reply:
> I guess (Default) can be explicitly deleted with PATCH(JSON-Patch), wheres (Required,Default) will get its Default value on delete?

I think that would make sense.

> On one hand PUT is used to "PUT exactly this resource there!", then we set some fields under its feet, and for (Default) fields we can loose ability to delete them.

Indeed:-) I think [this article](https://medium.freecodecamp.org/rest-is-the-new-soap-97ff6c09896d) says it in a good way:

> You want to use PUT to update your resource? OK, but some Holy Specifications state that the data input has to be equivalent to the representation received via a GET. So what do you do with the numerous read-only parameters returned by GET (creation time, last update time, server-generated token…)? You omit them and violate the PUT principles? You include them anyway, and expect an “HTTP 409 Conflict” if they don’t match server-side values (forcing you to then issue a GET...)? You give them random values and expect servers to ignore them (the joy of silent errors)? Pick your poison, REST clearly has no clue what a read-only attribute it, and this won’t be fixed anytime soon. Meanwhile, a GET is dangerously supposed to return the password (or credit card number) which was sent in a previous POST/PUT; good luck dealing with such write-only parameters too.

The thing is that no REST-API is able to completely follow the HTTP Standard and still be useful, and PUT is a particular tricky one. We actually used RPC calls in place of PUT in our own product, which allows us to design exactly the semantics we think makes sense for the app, and it allows us to do tailored things, like create multiple resources via one call...

But back to `PUT`, the truth is there is no 100% "correct" way to do it. Fields with default values and read-only fields just don't _fit_ with how PUT is  specified at all, but a REST API still need to set them in a way that's useful and at least follows some of the _principals_ of the HTTP standard while still retaining functionality.

Reply:
Thinking a bit more @smyrman about your "lights" example, I believe you can freely use (Required,Default) and it will work the way you want it. Isn't it?

Reply:
@rs Any input appreciated.

Reply:
>Thinking a bit more @smyrman about your "lights" example, I believe you can freely use (Required,Default) and it will work the way you want it. Isn't it?

Yes Setting all the fields with default to required would make it work.

>  EDIT: I am beginning to think rest-layer should not support (Default) alone. (Required) and (Required,Default) only.

Allowed or not, _PUT_ could/should _probably_ treat them equally...  Maybe it's a good idea to have a look at how JSON Schema is treating required fields with defaults. They have probably had a pretty similar discussion at some point...

@rs, yes your input is appreciated 🙂

Reply:
As an update, JSON Schema is extremely wage when [specifying](https://json-schema.org/latest/json-schema-validation.html#rfc.section.10.2) or [explaining](https://json-schema.org/understanding-json-schema/reference/generic.html#metadata) how default should work. I haven't found any JSON Schema discussions related to fields that are both required and has a default so far.

One thing to note though, is that JSON schema has `required` as a list of required properties in an `object`, and not as a property on each object field. The strength of having `requires` as a list, is that it allows setting `required` fields also for cases such as `schema.Dict`. I am considering this move in #77, but that just an FYI. For now, it is the way it is.

Reply:
I would really like to see it land, if you are able to. As for changes to do now, if you can add a description on how it works under different conditions to the PR, that would be the highest prority for me.

I think it would be nice to treat fields with `Default` equally to `Default` + `Required`, as the way things work would potentially not need to change for #77 later. Preferably I think this involves resting the field to Deefault, but if that makes e.g. JSON Patch not work as intended, then do what you think is best. The _most_ important part of this change, is that it no longer silently _clear_ required fields without erroring. For the rest of the PUT semantics, I actually don't mind to much, as long as the data corruption issue is resolved.

 > I am considering this move in #77, but that just an FYI. For now, it is the way it is.

This was just a note on the `Field.Required` property. Sorry if it was confusing.

Reply:
> This was just a note on the `Required` being a property of `Field`. Sorry if it was confusing.
No, its fine. 

So are there any additional code changes you may want, or it is just the README? I would really love to see (Default) go, and leave only (Required) and (Required,Default).
@rs your input on this matter is greatly appreciated as well.

I will refactor the tests for PUT, and will move each field type into separate `Test*` function.

Reply:
I am happy with just the README changes.